Initial Implementation

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR provides the initial implementation to create the following resources:
* S3 buckets in the Images (Production) and Images (Staging) accounts
* Roles providing full access to these buckets
* A user in the Users account
* Role allowing this user to assume the roles that provide full access to the S3 buckets
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This work provides the implementation to resolve https://github.com/cisagov/cool-system/issues/176
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I tested uploading, listing, and deleting an object in the buckets in both the Production and Staging Images accounts.

Images (Production):

```console
$ AWS_PROFILE=assessment-images-bucket-full-access-role-production aws s3 cp test_object.txt s3://cisa-cool-assessment-images-production
upload: ./test_object.txt to s3://cisa-cool-assessment-images-production/test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-production aws s3 ls s3://cisa-cool-assessment-images-production
2021-05-03 14:28:00          9 test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-production aws s3 rm s3://cisa-cool-assessment-images-production/test_object.txt
delete: s3://cisa-cool-assessment-images-production/test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-production aws s3 ls s3://cisa-cool-assessment-images-production
```

Images (Staging):

```console
$ AWS_PROFILE=assessment-images-bucket-full-access-role-staging aws s3 cp test_object.txt s3://cisa-cool-assessment-images-staging
upload: ./test_object.txt to s3://cisa-cool-assessment-images-staging/test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-staging aws s3 ls s3://cisa-cool-assessment-images-staging
2021-05-03 14:28:26          9 test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-staging aws s3 rm s3://cisa-cool-assessment-images-staging/test_object.txt
delete: s3://cisa-cool-assessment-images-staging/test_object.txt
$ AWS_PROFILE=assessment-images-bucket-full-access-role-staging aws s3 ls s3://cisa-cool-assessment-images-staging
```
<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

mcdonnnj committed:
Remove examples from repository

This is not a Terraform module so no examples are necessary.

mcdonnnj committed:
Create buckets and provide read access to Terraform state

This adds the Terraform code to create and attach the policies needed to create
the assessment images buckets, the Terraform code to create those buckets, and
the Terraform code to read the Terraform state for this project.

mcdonnnj committed:
Add the role and policies to provide full access to the buckets

Creates the polices that allow full access to their respective buckets and
attaches them to roles that will later be assumed by a user.

mcdonnnj committed:
Add user and policy to assume full bucket access roles

This adds the Terraform to create a user in the Users account and a policy that
allows it to assume the roles in the Images accounts that allow full access to
the assessment images buckets.

mcdonnnj committed:
Clean up tagging of resources

Make sure that appropriate resources are tagged. Add appropriate
production/staging tags to resources that will live in the Images (Production)
and Images (Staging) accounts.

mcdonnnj committed:
Update Terraform information in README

Use the more simplified `terraform-docs markdown --anchor=false .` command to
generate appropriate README contents.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Simply identifiers used in the project

Simplify the resource identifiers in this project to improve readability. I
have also adjusted a filename and two output names to better reflect what they
contain.

mcdonnnj committed:
Update variables and outputs

Update some variable names, descriptions, and defaults to improve readability
and consistency. Update some output descriptions for the same reasons.

mcdonnnj committed:
Update resources names for consistency

Update two resource names to be consistent with languages changes. Update a
missed reference to two of these resources. Fix a tyo in an input description.

jsf9k commented on assessment_images_bucket.tf:
```diff
@@ -0,0 +1,83 @@
+    aws_iam_role_policy_attachment.provision_bucket_production
+  ]
+
+  provider = aws.images_production
```
We've traditionally kept the keys that are generic to _all_ Terraform resource types in a separate section above the keys that are specific to the resource type.  This suggested change shows what I mean:
```suggestion
  provider = aws.images_production
  # Until this policy attachment happens, we don't have permission
  # to provision the bucket.
  depends_on = [
    aws_iam_role_policy_attachment.provision_bucket_production
  ]

  acl    = "private"
  bucket = local.production_bucket_name

```

I'm open to alphabetizing the first section (traditionally we have ordered them by perceived importance) but I think it makes sense to keep the two types of tags separate.  As always, I am open to discussion if the group feels this is wrong-headed.

FWIW, you seem to have followed the scheme I outlined above throughout the rest of this PR.

mcdonnnj replied:
Oops! Thanks for catching that. I have updated both resources in ca57fd3.

jsf9k commented on assessment_images_bucket.tf:
```diff
@@ -0,0 +1,83 @@
+    aws_iam_role_policy_attachment.provision_bucket_staging
+  ]
+
+  provider = aws.images_staging
```
See previous comment in this file.

jsf9k reviewed (changes requested):
Very strong work!

I had a point of discussion about the way you sorted the keys in a couple of resources.  Once that is resolved, either by sorting those keys as is done elsewhere in this PR or by the team deciding to change or sorting mechanism, I should be good to approve.

I am for:
* Keeping the tags specific to the resource type and the tags that are allowed on _any_ TF resource separate (this we already do)
* Sorting the tags that are not specific to the resource type alphabetically (this we do not yet do)

That said, I am fine with whatever the team decides.

mcdonnnj committed:
Reorganize aws_s3_bucket arguments

Correct and reorganize the arguments for the two aws_s3_bucket resources based
on feedback during review. Generally arguments that are universal should be at
the beginning of a resource while resource specific arguments should follow.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
Nice work!

dav3r commented on variables.tf:
```diff
@@ -4,36 +4,73 @@
-variable "ami_owner_account_id" {
+variable "assessment_images_bucket_base_name" {
+  type        = string
+  description = "The base name to use for the assessment images S3 buckets."
```
It might be worth clarifying exactly what is meant by "base name" here; perhaps add another sentence that shows the full names of the buckets with the default base name.

mcdonnnj replied:
Good point! I expanded on this description in c02ab6f. Please let me know if that is sufficient and understandable.

dav3r replied:
I like what you did- the only thing I'd change is to use our default bucket name "(e.g. `<base name>-production` for the bucket..." --> "(e.g. `cisa-cool-assessment-images-production` for the bucket..."

jsf9k replied:
I recently used [`bucket_prefix`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#bucket_prefix) instead of [`bucket`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#bucket) when configuring an S3 bucket via Terraform, admittedly for a use case where there will eventually be quite a few buckets.  One can still tell from the tags exactly what (in my case) account a bucket is associated with.

Just tossing that out there in case it's useful to someone else one day.

mcdonnnj replied:
Updated the description in a9503b2 (I rebased to correctly tie the README changes to the commit making the changes in a variable's information).

dav3r commented on variables.tf:
```diff
@@ -4,36 +4,73 @@
 
+variable "provisionaccount_role_name" {
+  type        = string
+  description = "The name of the IAM role that allows sufficient permissions to provision all AWS resources in the account."
```
I believe this is referring to the `Images` account here- if so, this variable can be replaced by pulling the role name from the appropriate remote state:
* `data.terraform_remote_state.images_production.outputs.provisionaccount_role.name`
* `data.terraform_remote_state.images_staging.outputs.provisionaccount_role.name`

mcdonnnj replied:
All of the COOL related repositories I looked at that used a `provisionaccount_role_name` variable so I was following suite. Do we use the remote state instead of a configurable variable in another project?

jsf9k replied:
@mcdonnnj, I've been following suit too but I think that method came into being before we started working with remote states.

dav3r replied:
My feeling is that whenever we can replace a variable with something from a remote state, it's worthwhile to do so.  Especially if it's a "core" remote state like `images_production` and `images_staging` here.

mcdonnnj replied:
This is a sensible argument and I made the change in 0edd170.

dav3r replied:
LGTM!

dav3r commented on providers.tf:
```diff
@@ -1,3 +1,46 @@
+}
+
+# The provider used to create the role that can be assumed to do
+# everything the CI user needs to do in the production account.
```
```suggestion
# everything the CI user needs to do in the Images (Production) account.
```

mcdonnnj replied:
I cleaned up this comment in c02ab6f. Please let me know if you have further feedback.

dav3r commented on providers.tf:
```diff
@@ -1,3 +1,46 @@
+}
+
+# The provider used to create the role that can be assumed to do
+# everything the CI user needs to do in the staging account.
```
```suggestion
# everything the CI user needs to do in the Images (Staging) account.
```

mcdonnnj replied:
Also cleaned up in c02ab6f.

dav3r commented on provisionassessmentimagesbucket_policy.tf:
```diff
@@ -0,0 +1,44 @@
+data "aws_iam_policy_document" "provision_bucket_production" {
+  statement {
+    actions = [
+      "s3:*",
```
Typically, we prefer to specify all of the permissions that are needed, rather than wildcard it.  This makes it explicit which permissions have been granted and also forces us to document when new permissions are added/removed.

mcdonnnj replied:
I can adjust it, but I was following the way it was done in [cisagov/cool-dns-certboto](https://github.com/cisagov/cool-dns-certboto/blob/develop/provisioncertificatesbucket_policy.tf#L6-L18). Although it is a wildcard action, it is limited to a specific bucket per https://github.com/cisagov/cool-images-assessment-images/blob/ca57fd308985405439247e84f34d99ebbe42d487/provisionassessmentimagesbucket_policy.tf#L12-L14

jsf9k replied:
I think since he is creating the bucket and limiting the resource the `s3:*` is OK here.  This is how I have been doing it, so I'll need to make some corrections if we decide we don't like this.

dav3r replied:
Even when it's limited to a bucket created here, I still think there's benefit in specifying the permissions, rather than using a wildcard.  I don't think there are all that many S3 permissions needed here, are there?  I'm ok with leaving the wildcard if you guys are, but I'd prefer to see us working towards eliminating the `*`s where we can.

mcdonnnj replied:
I have opened https://github.com/cisagov/cool-images-assessment-images/issues/2 to narrow the permissions on this policy.

dav3r commented on assessmentimagesbucketfullaccess_policy.tf:
```diff
@@ -0,0 +1,64 @@
+# ------------------------------------------------------------------------------
+# Create the IAM policies that allows full access to the assessment images
```
```suggestion
# Create the IAM policies that allow full access to the assessment images
```

mcdonnnj replied:
Thanks for catching these grammatical errors. Please see 5749084.

dav3r commented on assessmentimagesbucketfullaccess_role.tf:
```diff
@@ -0,0 +1,36 @@
+# ------------------------------------------------------------------------------
+# Create the IAM role that allows full access to the assessment images
```
```suggestion
# Create the IAM roles that allow full access to the assessment images
```

mcdonnnj replied:
Please see 5749084.

dav3r commented on read_terraform_state_role.tf:
```diff
@@ -0,0 +1,26 @@
+  # production or vice versa, which is highly annoying.
+  role_tags                   = { for k, v in var.tags : k => v if k != "Workspace" }
+  terraform_state_bucket_name = "cisa-cool-terraform-state"
+  terraform_state_path        = "cool-images-assessment-images/*.tfstate"
```
Can we just make this `cool-images-assessment-images/terraform.tfstate`?  Or is there some benefit to using `*.tfstate` here that I'm not thinking of?
```suggestion
  terraform_state_path        = "cool-images-assessment-images/terraform.tfstate"
```

mcdonnnj replied:
The other COOL repositories I looked at do the same thing so I was following suite. I can change it, but I'd like clarity on when it should be a wildcard and when it shouldn't.

jsf9k replied:
This _probably_ started because the first repo from which we read remote state was [cisagov/cool-accounts](https://github.com/cisagov/cool-accounts), where this sort of voodoo would make sense.

Eventually even [cisagov/cool-accounts](https://github.com/cisagov/cool-accounts) will be split into separate repos, at which time the wildcard will no longer make sense even for it.

We definitely need to get some of this stuff written down in the style guide.  I often don't remember the better way of doing things and just copy what we did in the distant path.

dav3r replied:
I think most of our use cases can avoid using `*.tfstate` since we very rarely have different TF state objects in the same subdirectory.  I think the `*.tfstate` should only be used when there is a chance that multiple TF states will be in use- but we tend to not do that anywhere.

mcdonnnj replied:
After discussion this makes sense. I've limited the scope in a85157a.

dav3r replied:
Spiffy.

dav3r commented on variables.tf:
```diff
@@ -4,36 +4,73 @@
+
+variable "provisionassessmentimagesbucket_policy_description" {
+  type        = string
+  description = "The description to associate with the IAM policy that allows provisioning of the assessment images S3 Bucket in the Images account."
```
No need to capitalize "bucket" here.
```suggestion
  description = "The description to associate with the IAM policy that allows provisioning of the assessment images S3 bucket in the Images account."
```

mcdonnnj replied:
Please see 5749084.

dav3r commented on variables.tf:
```diff
@@ -4,36 +4,73 @@
+
+variable "provisionassessmentimagesbucket_policy_name" {
+  type        = string
+  description = "The name to associate with the IAM policy that allows provisioning of the assessment images S3 Bucket in the Images account."
```
See previous comment.
```suggestion
  description = "The name to associate with the IAM policy that allows provisioning of the assessment images S3 bucket in the Images account."
```

mcdonnnj replied:
Please see 5749084.

dav3r commented on variables.tf:
```diff
@@ -4,36 +4,73 @@
+
+variable "user_name" {
+  type        = string
+  description = "The name of the user to create in the Users account which will have the attached policy to assume roles allowing access to the assessment images buckets in the Images (Production) and Images (Staging) accounts."
```
To me this reads more cleanly, but take it or leave it:
```suggestion
  description = "The name of the user to create in the Users account that can assume roles allowing access to the assessment images buckets in the Images (Production) and Images (Staging) accounts."
```

mcdonnnj replied:
I tweaked this based on your feedback in c02ab6f. Please let me know if you think it need further refinement.

dav3r replied:
Looks great to me. 👍

dav3r reviewed (changes requested):
Excellent work here!  I noted a handful of minor thangs (mostly grammar in comments) and one item that definitely should be addressed (replacing a variable with something that can be obtained from a remote state).

mcdonnnj committed:
Fix grammatical errors

Fix some grammatical mistakes that were noticed in review. Update the README
to reflect these updates.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Update language in comments and variable descriptions

Improve hte readability of some comments and a variable description. Expand on
the description for a variable to better explain how it is used. Update the
README to reflect the changes in the variable descriptions.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Switch to using remote state to determine a role name

Instead of having a `provisionaccount_role_name` variable we can use the output
from the appropriate remote state to get the role name. This will reduce input
complexity and allow situations where the two accounts might have used
different role names for the role serving the same purpose.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Narrow the resources for Terraform remote state access

It was pointed out that a wildcard was likely inappropriate as a scope for
tfstate files in the remote state. Except for the rare exception we would only
need access to `terraform.tfstate`. Since this project is _not_ one of those
exceptions we have limited the scope of the permission.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>
Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r reviewed (approved):
Everything looks great now - nice work! 🧨
